File: app/fsharp/bolero/src/Domain/Board.fs
module PuyoGame.Domain.Board

open PuyoGame.Domain

/// ボード
type Board = private {
    Width: int
    Height: int
    Cells: Cell array array
}

let width (board: Board) = board.Width
let height (board: Board) = board.Height
let getAllCells (board: Board) = board.Cells

/// 空のゲームボードを作成
let createEmpty : Board =
    {
        Width = 8
        Height = 12
        Cells = Array.init 12 (fun _ -> Array.create 8 Empty)
    }

/// 座標が有効な範囲内かチェック
let isValidPosition (board: Board) (pos: Position) : bool =
    pos.X >= 0 && pos.X < board.Width && pos.Y >= 0 && pos.Y < board.Height

/// 指定位置のセルを取得
let tryGetCell (board: Board) (pos: Position) : Cell option =
    if isValidPosition board pos then
        Some board.Cells.[pos.Y].[pos.X]
    else
        None

/// 指定位置にぷよがあるかチェック
let hasPuyo (board: Board) (pos: Position) : bool =
    match tryGetCell board pos with
    | Some (Filled _) -> true
    | _ -> false

/// 指定位置のセルを更新（不変）
let setCell (board: Board) (pos: Position) (cell: Cell) : Board =
    if not (isValidPosition board pos) then
        board
    else
        let newCells = Array.copy board.Cells
        newCells.[pos.Y] <- Array.copy newCells.[pos.Y]
        newCells.[pos.Y].[pos.X] <- cell
        { board with Cells = newCells }

/// 全消し判定
let isPerfectClear (board: Board) : bool =
    board.Cells
    |> Array.forall (Array.forall ((=) Empty))

File: app/scala/src/main/scala/puyo/Board.scala
package puyo

// ========== ゲームボード ==========
case class Board(cells: Vector[Vector[Int]]) {
  val height: Int = cells.size
  val width: Int = cells.headOption.map(_.size).getOrElse(0)

  def isValidPosition(x: Int, y: Int): Boolean =
    x >= 0 && x < width && y >= 0 && y < height

  def getPuyoAt(x: Int, y: Int): Option[Int] =
    if (isValidPosition(x, y)) Some(cells(y)(x))
    else None

  def setPuyo(x: Int, y: Int, color: Int): Board =
    if (isValidPosition(x, y))
      copy(cells = cells.updated(y, cells(y).updated(x, color)))
    else this

  def isPerfectClear: Boolean =
    cells.flatten.forall(_ == 0)
}

object Board {
  val BoardWidth = 8
  val BoardHeight = 12

  def createEmpty: Board =
    Board(Vector.fill(BoardHeight, BoardWidth)(0))
}
